### PR TITLE
feat: add multiply/divide functions

### PR DESCRIPTION
**Pull request by EscapedGibbon**

### CODE REVIEW

**Comment by EscapedGibbon** [comment]
I haven't added functions as image methods. Also right now functions do not multiply alpha channels, maybe it is not needed at all, or maybe we can add an option if alpha channel should be taken into account as well.

File: src/compute/multiply.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  image: Image,
+  value: number,
+  options: MultiplyOptions = {},
+) {
+  const {
+    channels = new Array(image.components).fill(0).map((value, index) => index),
+  } = options;
+  validateChannels(channels, image);
+
+  const newImage = getOutputImage(image, options, { clone: true });
```
Comment (stropitek) [inline src/compute/multiply.ts:35]: Why didn't you provide the `out` option which let's you choose where to put the result, in the same way we do in many other processing methods?

File: src/compute/divide.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  image: Image,
+  value: number,
+  options: MultiplyOptions = {},
+) {
+  const {
+    channels = new Array(image.components).fill(0).map((value, index) => index),
+  } = options;
+  validateChannels(channels, image);
+  if (value === 0) {
+    throw new Error(`Value cannot be equal to 0.`);
```
Comment (stropitek) [inline src/compute/divide.ts:26]:
```suggestion
    throw new TypeError(`Cannot divide by 0`);
```

1. Use TypeError when you receive a bad argument
2. We don't put periods in our error messages.
3. Try to write the message in a way that the error is easy to interpret by the user.

File: src/compute/__tests__/divide.test.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  ]);
+  image = divide(image, 2);
+  const result = testUtils.createRgbaImage([
+    [115, 40, 60, 255],
+    [50, 70, 6, 1],
+  ]);
+  expect(image).toStrictEqual(result);
+});
+
+test('must throw error', () => {
```
Comment (stropitek) [inline src/compute/__tests__/divide.test.ts:16]:
```suggestion
test('error when dividing by 0', () => {
```

The name of the test is too generic

File: src/compute/multiply.ts (hunk 1 above)
Comment (EscapedGibbon) [reply src/compute/multiply.ts:35]: I looked at several functions as examples(level, invert, pixelate) so I did it the same way. What did I do wrong?

File: src/compute/divide.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+//The used options are the same as in multiply function.
+import { MultiplyOptions } from './multiply';
+
+/**
+ *
+ * Divides image pixels by a certain value.
+ * @param image - image to which division will be applied.
+ * @param value - Value by which each pixel will be divided.
+ * @param options - Multiply options
```
Comment (stropitek) [inline src/compute/divide.ts:13]: Divide not multiply

File: src/compute/multiply.ts (hunk 1 above)
Comment (stropitek) [reply src/compute/multiply.ts:35]: Sorry I missed something while reviewing

File: src/compute/multiply.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  if (channels.length === 0) {
+    return newImage;
+  }
+  for (let channel = 0; channel < channels.length; channel++) {
```
Comment (stropitek) [inline src/compute/multiply.ts:39]: It's a bit confusing to name this `channel` when it is actually the index in which to look for the channel.

File: src/compute/multiply.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    for (let row = 0; row < newImage.height; row++) {
+      for (let column = 0; column < newImage.width; column++) {
+        const newIntensity =
+          newImage.getValue(column, row, channels[channel]) * value;
+        newImage.setClampedValue(column, row, channel, newIntensity);
```
Comment (stropitek) [inline src/compute/multiply.ts:44]: isn't the third parameter wrong here?

**Comment by stropitek** [comment]
> Also right now functions do not multiply alpha channels

Is it not possible by providing the alpha channel in the `channels` option?

**Comment by stropitek** [comment]
We need at least one test to test each option (`channels` and `out`)

**Comment by stropitek** [comment]
You can also add `divide` and `multiply` as `Image` methods

File: src/compute/divide.ts (hunk 4 above)
Comment (EscapedGibbon) [reply src/compute/divide.ts:13]: It should be, but the interface I use is from MultiplyOptions. In the parameters docs should I put that it is divide options anyway?

Comment (targos) [reply src/compute/divide.ts:13]: You should not use MultiplyOptions.

Comment (EscapedGibbon) [reply src/compute/divide.ts:13]: But if the options are the same as with multiply function, should I create a DivideOptions interface anyway? Or should I do it differently?

**Comment by targos** [comment]
@stropitek Was it discussed to put these methods in the `compute` folder? That doesn't seem right to me as it doesn't contain similar methods.
I think `add` and `subtract` are similar.

File: src/compute/divide.ts (hunk 4 above)
Comment (stropitek) [reply src/compute/divide.ts:13]: It's fine to have some duplicated types.

If you don't want to duplicate types, you can always do something like this:
```ts
interface OperationOptionsBase {
  channels?: number[];
  out?: Image
}

interface  MultiplyOptions extends OperationOptionsBase {};
interface  DivideOptions extends OperationOptionsBase {};
```

Comment (targos) [reply src/compute/divide.ts:13]: Yes, in general, 1 function with options = 1 interface

**Comment by stropitek** [comment]
> Was it discussed to put these methods in the `compute` folder? That doesn't seem right to me as it doesn't contain similar methods.

No this wasn't discussed and I did not pay attention to that during the review.

> I think `add` and `subtract` are similar.

I agree

File: src/compare/multiply.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  /**
+   * Channels where value will be multiplied.
+   * @default all channels
+   */
+  channels?: number[];
+  /**
+   * Image to which the resulting image has to be put.
+   */
+  out?: Image;
+}
```
Comment (stropitek) [inline src/compare/multiply.ts:15]: This is not good.

First the name is not good. Then you use the term "multiply" to document something that is used by multiply and divide.

There are many other methods that use those 2 options. To do this properly, the base should be shared with many other methods (so big refactoring).

So please just use 2 types `MultiplyOptions` and `DivideOptions`. It's ok that they have duplicate properties.

**Review by stropitek — CHANGES_REQUESTED** [review]

**Review by stropitek — APPROVED** [review]

**Pull request merged**